### ISSUE TITLE
PyTorch 基础

### ISSUE THREAD

**Issue by zhaoyiyidan** [zhaoyiyidan/recordstudy#30]
以下是 PyTorch 中常用的核心 API 分类整理，涵盖张量操作、模型构建、训练流程等关键模块。通过示例代码帮助快速掌握用法。

---

### **1. 张量操作（Tensor Operations）**
PyTorch 的核心数据结构，支持数学运算、形状变换、索引等。

#### **创建张量**
```python
# 基础创建
x = torch.tensor([1, 2, 3])          # 从列表创建
y = torch.rand(3, 3)                 # 均匀分布随机张量 (0~1)
z = torch.randn(3, 3)                # 标准正态分布随机张量
a = torch.zeros(2, 2)                # 全零张量
b = torch.ones(3)                    # 全一张量
c = torch.arange(0, 10, 2)           # 类似 range：0, 2, 4, 6, 8

# 指定设备（GPU）
device = torch.device("cuda")
gpu_tensor = torch.ones(3, 3, device=device)
```

#### **形状操作**
```python
x = torch.randn(4, 3, 28, 28)        # 4张3通道的28x28图像
x.view(4, 3, 784)                    # 改变形状（总元素数需一致）
x.reshape(4, -1)                     # 自动推导维度（-1占位）
x.permute(0, 2, 3, 1)               # 调整维度顺序（如 NHWC 格式）
x.squeeze(dim=1)                    # 压缩维度（删除大小为1的维度）
x.unsqueeze(dim=0)                  # 扩展维度（增加大小为1的维度）
```

#### **数学运算**
```python
a + b                               # 逐元素加法（等价于 torch.add(a, b)）
a @ b                               # 矩阵乘法（等价于 torch.matmul(a, b)）
torch.sum(x, dim=0)                 # 沿维度求和
torch.mean(x, dim=1)                # 沿维度求均值
torch.max(x, dim=2, keepdim=True)   # 沿维度求最大值，保留维度
torch.exp(x)                        # 指数运算
torch.clamp(x, min=0, max=1)       # 裁剪值到 [0,1] 区间（类似 ReLU）
```

#### **索引与切片**
```python
x[:, 1:3, ::2, :]                   # 类似 NumPy 的多维切片
mask = x > 0.5                      # 布尔掩码索引
selected = torch.where(mask, x, 0)  # 条件选择（满足mask取x，否则取0）
indices = torch.tensor([0, 2])
x.index_select(dim=1, indices)      # 按索引选择特定维度数据
```

---

### **2. 神经网络层（`torch.nn` 模块）**
预定义的神经网络层和工具。

#### **常用层**
```python
import torch.nn as nn

# 全连接层
fc = nn.Linear(in_features=784, out_features=256)

# 卷积层（2D卷积）
conv = nn.Conv2d(in_channels=3, out_channels=64, kernel_size=3, stride=1, padding=1)

# 循环神经网络层（LSTM）
lstm = nn.LSTM(input_size=100, hidden_size=256, num_layers=2, batch_first=True)

# 归一化层
batchnorm = nn.BatchNorm2d(num_features=64)
layernorm = nn.LayerNorm(normalized_shape=256)

# 激活函数
relu = nn.ReLU()
sigmoid = nn.Sigmoid()
softmax = nn.Softmax(dim=1)
```

#### **容器（模型组装）**
```python
# 顺序容器（按顺序执行层）
model = nn.Sequential(
    nn.Conv2d(3, 64, 3),
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Flatten(),
    nn.Linear(64 * 13 * 13, 10)
)

# 自定义模型（继承 nn.Module）
class CNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 64, 3)
        self.pool = nn.MaxPool2d(2)
        self.fc = nn.Linear(64 * 13 * 13, 10)
  
    def forward(self, x):
        x = self.pool(nn.ReLU()(self.conv1(x)))
        x = x.view(x.size(0), -1)  # 展平
        return self.fc(x)
```

---

### **3. 损失函数（Loss Functions）**
```python
# 分类任务
loss_fn = nn.CrossEntropyLoss()     # 交叉熵损失（搭配 Logits 输入）
loss_fn = nn.BCEWithLogitsLoss()    # 二分类交叉熵（输入未归一化）

# 回归任务
loss_fn = nn.MSELoss()              # 均方误差
loss_fn = nn.L1Loss()               # 平均绝对误差

# 自定义损失函数
def my_loss(output, target):
    return torch.mean((output - target) ** 2)
```

---

### **4. 优化器（Optimizers）**
```python
from torch.optim import SGD, Adam, RMSprop

# 随机梯度下降
optimizer = SGD(model.parameters(), lr=0.01, momentum=0.9)

# 自适应优化器（常用）
optimizer = Adam(model.parameters(), lr=1e-3, weight_decay=1e-4)  # 带 L2 正则化

# 学习率调度器（配合优化器使用）
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=10, gamma=0.1)
```

---

### **5. 数据加载（`DataLoader`）**
```python
from torch.utils.data import DataLoader, Dataset, TensorDataset

# 自定义数据集
class MyDataset(Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels
  
    def __len__(self):
        return len(self.data)
  
    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]

# 创建 DataLoader
dataset = MyDataset(X, y)
dataloader = DataLoader(dataset, batch_size=32, shuffle=True, num_workers=4)

# 快速封装（若数据已是张量）
tensor_dataset = TensorDataset(features_tensor, labels_tensor)
```

---

### **6. 设备管理（GPU/CPU）**
```python
# 检测 GPU 是否可用
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# 移动张量/模型到设备
x = x.to(device)
model = model.to(device)

# 释放 GPU 内存
torch.cuda.empty_cache()

# 多 GPU 并行训练
if torch.cuda.device_count() > 1:
    model = nn.DataParallel(model)
```

---

### **7. 模型保存与加载**
```python
# 保存模型参数
torch.save(model.state_dict(), "model.pth")

# 加载模型参数
model.load_state_dict(torch.load("model.pth"))

# 保存整个模型（包含结构）
torch.save(model, "full_model.pth")
loaded_model = torch.load("full_model.pth")
```

---

### **8. 训练流程关键 API**
```python
# 前向传播
outputs = model(inputs)

# 计算损失
loss = loss_fn(outputs, labels)

# 反向传播
loss.backward()                     # 自动计算梯度

# 参数更新
optimizer.step()                    # 更新参数
optimizer.zero_grad()               # 清空梯度（防止累积）

# 关闭梯度计算（推理/评估时）
with torch.no_grad():
    predictions = model(test_inputs)
```

---

### **常用工具库**
- **`torchvision`**: 图像数据处理（数据集、预处理）
  ```python
  from torchvision import transforms, datasets

  transform = transforms.Compose([
      transforms.Resize(224),
      transforms.ToTensor(),
      transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
  ])
  dataset = datasets.CIFAR10(root="./data", transform=transform, download=True)
  ```
- **`torchtext`**: 文本数据处理
- **`torchaudio`**: 音频数据处理

---

掌握这些 API 后，可以快速搭建和训练深度学习模型。建议结合官方文档 ([pytorch.org/docs](https://pytorch.org/docs/stable/index.html)) 深入学习更高级的功能（如分布式训练、量化部署）。
